RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc

```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+
+  for (unsigned i = 0; i < trackstersTRK.size(); ++i) {
+    auto mergedIdx = indexInMergedCollTRK[i];
+    auto &t = trackstersTRK[i];  //trackster
+
+    if (!usedTrackstersMerged[mergedIdx] and t.raw_energy() > 0) {
+      auto trackIdx = t.seedIndex();
+      auto const &track = tracks[trackIdx];
+      if (!usedSeeds[trackIdx]) {
+        usedSeeds[trackIdx] = true;
+        usedTrackstersMerged[mergedIdx] = true;
+        float p = t.raw_pt() * cosh(t.barycenter().eta());
+        float energy = std::sqrt(p * p + mpion2);
+        TICLCandidate tmpCandidate;
+        tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, mergedIdx));
+        tmpCandidate.setCharge(track.charge());
+        tmpCandidate.setTrackPtr(edm::Ptr<reco::Track>(track_h, trackIdx));
+        tmpCandidate.setPdgId(211 * track.charge());
+        tmpCandidate.setRawEnergy(energy);
+        math::XYZTLorentzVector p4(
+            p * track.momentum().unit().x(), p * track.momentum().unit().y(), p * track.momentum().unit().z(), energy);
+        tmpCandidate.setP4(p4);
+        resultCandidates->push_back(tmpCandidate);
+      }
     }
-    tracksterEM_idx++;
   }
-
-  tracksterHAD_idx = 0;
-  for (auto const &t : trackstersHAD) {
-    if (debug_) {
-      LogDebug("TrackstersMergeProducer") << " Considering trackster " << tracksterHAD_idx
-                                          << " as used: " << usedTrackstersHAD[tracksterHAD_idx] << std::endl;
+  // For all seeds that have 0-energy tracksters whose track is not marked as used, create a charged hadron with the track information.
+  for (auto const &s : seedingTrk) {
+    if (usedSeeds[s.index] == false) {
+      auto const &track = tracks[s.index];
+      // emit a charged hadron
+      TICLCandidate tmpCandidate;
+      tmpCandidate.setCharge(track.charge());
+      tmpCandidate.setTrackPtr(edm::Ptr<reco::Track>(track_h, s.index));
+      tmpCandidate.setPdgId(211 * track.charge());
+      float energy = std::sqrt(track.pt() * track.pt() + mpion2);
+      tmpCandidate.setRawEnergy(energy);
+      math::XYZTLorentzVector p4(track.momentum().x(), track.momentum().y(), track.momentum().z(), energy);
+      tmpCandidate.setP4(p4);
+      resultCandidates->push_back(tmpCandidate);
+      usedSeeds[s.index] = true;
     }
-    if (!usedTrackstersHAD[tracksterHAD_idx]) {
-      LogDebug("TrackstersMergeProducer")
-          << " Creating a neutral hadron from HAD Trackster with track energy " << t.raw_energy() << " and direction "
-          << t.eigenvectors(0).eta() << ", " << t.eigenvectors(0).phi() << std::endl;
-      result->push_back(t);
+  }
+  // for all general tracks (high purity, pt > 1), check if they have been used: if not, promote them as charged hadrons
+  for (unsigned i = 0; i < tracks.size(); ++i) {
+    auto const &track = tracks[i];
+    if (track.pt() > 1.f and track.quality(reco::TrackBase::highPurity) and track.missingOuterHits() < 5 and
```

Comment:
a lot of magic numbers here: should be named constants, configurable, or maybe obtained from geometry